nni/compression/pytorch/utils/counter.py

```diff
@@ -1,18 +1,217 @@
+    }
+
+    add_results(m._name, **results)
+
+def count_bn(m, x, y, add_results):
+    results = {
+        "flops": 2 * x[0].numel(),
+        "params": _get_params(m),
+        "weight_size": tuple(m.weight.size()),
+        "input_size": tuple(x[0].size()),
+        "output_size": tuple(y.size()),
+        "module_type": type(m).__name__
+    }
+
+    add_results(m._name, **results)
+
+
+def count_relu(m, x, y, add_results):
+    results = {
+        "flops": x[0].numel(),
+        "params": 0,
+        "weight_size": tuple(m.weight.size()),
+        "input_size": tuple(x[0].size()),
+        "output_size": tuple(y.size()),
+        "module_type": type(m).__name__
+    }
+
+    add_results(m._name, **results)
+
+def count_avgpool(m, x, y, add_results):
+    print("avgpool", y.numel())
+    results = {
+        "flops": y.numel(),
+        "params": 0,
+        "weight_size": 0,
+        "input_size": tuple(x[0].size()),
+        "output_size": tuple(y.size()),
+        "module_type": type(m).__name__
+    }
+
+    add_results(m._name, **results)
+
+
+def count_adap_avgpool(m, x, y, add_results):
+    kernel = torch.DoubleTensor([*(x[0].shape[2:])]) // torch.DoubleTensor(list((m.output_size,))).squeeze()
+    total_add = torch.prod(kernel)
+    total_div = 1
+    kernel_ops = total_add + total_div
+    num_elements = y.numel()
+
+    results = {
+        "flops": kernel_ops * num_elements,
+        "params": 0,
+        "weight_size": 0,
+        "input_size": tuple(x[0].size()),
+        "output_size": tuple(y.size()),
+        "module_type": type(m).__name__
+    }
+
+    add_results(m._name, **results)
+
+
+def count_upsample(m, x, y, add_results):
+    if m.mode not in ("linear", "bilinear", "bicubic"):
+        return
+
+    if m.mode == "linear":
+        total_ops = y.nelement() * 5  # 2 muls + 3 add
+    elif m.mode == "bilinear":
+        # https://en.wikipedia.org/wiki/Bilinear_interpolation
+        total_ops = y.nelement() * 11  # 6 muls + 5 adds
+    elif m.mode == "bicubic":
+        # https://en.wikipedia.org/wiki/Bicubic_interpolation
+        # Product matrix [4x4] x [4x4] x [4x4]
+        ops_solve_A = 224  # 128 muls + 96 adds
+        ops_solve_p = 35  # 16 muls + 12 adds + 4 muls + 3 adds
+        total_ops = y.nelement() * (ops_solve_A + ops_solve_p)
+    elif m.mode == "trilinear":
+        # https://en.wikipedia.org/wiki/Trilinear_interpolation
+        # can viewed as 2 bilinear + 1 linear
+        total_ops = y.nelement() * (13 * 2 + 5)
+
+    results = {
+        "flops": total_ops,
+        "params": 0,
+        "weight_size": tuple(m.weight.size()),
+        "input_size": tuple(x[0].size()),
+        "output_size": tuple(y.size()),
+        "module_type": type(m).__name__
+    }
+
+    add_results(m._name, **results)
+
+def count_naive(m, _, __, add_results):
+    pass
+
+DEFAULT_OPS = {
+    nn.Conv1d: count_convNd,
+    nn.Conv2d: count_convNd,
+    nn.Conv3d: count_convNd,
+    nn.Linear: count_linear
```

Comment:
why `DEFAULT_OPS` only includes convNd and linear?

Reply:
As far as I know, most papers only count flops of Conv and Linear~